xlators/cluster/dht/src/dht-inode-write.c

```diff
@@ -218,211 +345,151 @@ dht_writev(call_frame_t *frame, xlator_t *this, fd_t *fd, struct iovec *vector,
     local->rebalance.iobref = iobref_ref(iobref);
     local->call_cnt = 1;
 
-    STACK_WIND_COOKIE(frame, dht_writev_cbk, subvol, subvol,
-                      subvol->fops->writev, fd, local->rebalance.vector,
-                      local->rebalance.count, local->rebalance.offset,
-                      local->rebalance.flags, local->rebalance.iobref,
-                      local->xattr_req);
+    dht_inode_write_wind(frame, subvol);
 
     return 0;
 
 err:
     op_errno = (op_errno == -1) ? errno : op_errno;
-    DHT_STACK_UNWIND(writev, frame, -1, op_errno, NULL, NULL, NULL);
+    dht_inode_write_unwind(GF_FOP_WRITE, frame, -1, op_errno, NULL, NULL, NULL);
 
     return 0;
 }
 
 int
-dht_truncate_cbk(call_frame_t *frame, void *cookie, xlator_t *this, int op_ret,
-                 int op_errno, struct iatt *prebuf, struct iatt *postbuf,
-                 dict_t *xdata)
+dht_truncate(call_frame_t *frame, xlator_t *this, loc_t *loc, off_t offset,
+             dict_t *xdata)
 {
+    xlator_t *subvol = NULL;
+    int op_errno = -1;
     dht_local_t *local = NULL;
-    xlator_t *prev = NULL;
-    int ret = -1;
-    xlator_t *src_subvol = NULL;
-    xlator_t *dst_subvol = NULL;
-    inode_t *inode = NULL;
-
-    GF_VALIDATE_OR_GOTO("dht", frame, err);
-    GF_VALIDATE_OR_GOTO("dht", this, out);
-    GF_VALIDATE_OR_GOTO("dht", frame->local, out);
-    GF_VALIDATE_OR_GOTO("dht", cookie, out);
-
-    local = frame->local;
-    prev = cookie;
-
-    /* Needs to be checked only for ftruncate.
-     * ftruncate fails with EBADF/EINVAL if dht has not yet opened the fd
-     * on the cached subvol. This could happen if the file was migrated
-     * and a lookup updated the cached subvol in the inode ctx.
-     * We only check once as this could actually be a valid error.
-     */
-
-    if ((local->fop == GF_FOP_FTRUNCATE) &&
-        dht_check_remote_fd_failed_error(local, op_ret, op_errno)) {
-        ret = dht_check_and_open_fd_on_subvol(this, frame);
-        if (ret)
-            goto out;
-        return 0;
-    }
-
-    if ((op_ret == -1) && !dht_inode_missing(op_errno)) {
-        local->op_errno = op_errno;
-        local->op_ret = -1;
-        gf_msg_debug(this->name, op_errno, "subvolume %s returned -1",
-                     prev->name);
 
-        goto out;
-    }
+    VALIDATE_OR_GOTO(frame, err);
```

Comment:
nit: might be an opportunity to get rid of validations that are not really needed? @amarts  commented on those in the past, that they are not needed, as you'd crash in the wind/unwind macros if frame or this are NULL.

Reply:
Actually neither of these are needed. There are bigger problems if they are NULL. Let them be exposed and crash the process so we find them sooner in dev process itself. In AFR no such validations happen so far no crashes are seen. I will remove them.